File: PixelSolution/PixelEngine/UIText.cpp
#include "pch.h"
#include "UIText.h"
#include "PixelEngineAPI.h"
#include "Registry.h"
#include "Graphics.h"
#include <algorithm>
#include <cstring>
#include "PixelEngine.h"
#include "ResourceManager.h"
extern PixelEngine* Engine;
UITextData* UIText_Add(unsigned int id)
{
	auto registry = GetRegistry();
	registry->Add<UITextData>(id);
	if (registry->Has<GraphicsData>(id) == false)
	{
		registry->Add<GraphicsData>(id);
	}
	auto data = registry->Get<UITextData>(id);
	auto graphicsData = registry->Get<GraphicsData>(id);
	data->thisID = id;
	graphicsData->renderingData.renderType = RENDER_TYPE::TEXT;
	graphicsData->renderingData.passType = PASS_TYPE::UI;
	graphicsData->renderingData.text.color[0] = 1.0f;
	graphicsData->renderingData.text.color[1] = 1.0f;
	graphicsData->renderingData.text.color[2] = 1.0f;
	graphicsData->renderingData.text.color[3] = 1.0f;
	return data;
}

UITextData* UIText_Get(unsigned int id)
{
	auto registry = GetRegistry();
	UITextData* data = registry->Get<UITextData>(id);
	if (data == nullptr)
	{
		PixelLog::Error("[UIText][GetComponent] Not Find Component");
	}
	return data;
}

bool UIText_Has(unsigned int id)
{
	auto registry = GetRegistry();
	return registry->Get<UITextData>(id) != nullptr;
}

void UIText_SetText(unsigned int id, const char* text)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<UITextData>(id);
	auto graphics = registry->Get<GraphicsData>(id);
	if (data1 != nullptr && graphics != nullptr)
	{
		auto& target = graphics->renderingData.text;
		std::memset(target.content, 0, sizeof(target.content));

		const std::string textSTR = text != nullptr ? text : "";
		const size_t length = (std::min)(textSTR.size(), sizeof(target.content) - 1);
		if (length > 0)
		{
			std::memcpy(target.content, textSTR.data(), length);
		}

		data1->text = textSTR;
	}
	else
	{
		PixelLog::Error("[UIText][SetText] Not Find Component");
	}
}

void UIText_SetTextSize(unsigned int id, float scale)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<UITextData>(id);
	auto graphics = registry->Get<GraphicsData>(id);
	if (data1 != nullptr)
	{
		graphics->renderingData.text.scale = scale;
	}
	else
	{
		PixelLog::Error("[Renderer2D][SetTextSize] Not Find Component");
	}
}

void UIText_SetFont(unsigned int id, const char* name)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<UITextData>(id);
	auto graphics = registry->Get<GraphicsData>(id);
	if (data1 != nullptr)
	{
		std::string fontName(name);
		auto textureID = Engine->GetResourceID(RESOURCE_TYPE::FONT, fontName);
		graphics->renderingData.text.fontKey = textureID;
	}
	else
	{
		PixelLog::Error("[Renderer2D][SetFont] Not Find Component");
	}
}

void UIText_SetColor(unsigned int id, float r, float g, float b, float a)
{
	auto registry = GetRegistry();
	auto textData = registry->Get<UITextData>(id);
	auto graphics = registry->Get<GraphicsData>(id);
	if (textData == nullptr || graphics == nullptr)
	{
		PixelLog::Error("[UIText][SetColor] Not Find Component");
		return;
	}

	auto normalizeColor = [](float value)
		{
			return (std::clamp)(value, 0.0f, 255.0f) / 255.0f;
		};

	graphics->renderingData.text.color[0] = normalizeColor(r);
	graphics->renderingData.text.color[1] = normalizeColor(g);
	graphics->renderingData.text.color[2] = normalizeColor(b);
	graphics->renderingData.text.color[3] = normalizeColor(a);
}

File: Survival/Engine/UIText.lua
local ffi = require("ffi")
local dll = ffi.load("PixelEngine")

ffi.cdef[[
	typedef struct {
		unsigned int thisID;
	} UITextData;

	UITextData* UIText_Add(unsigned int ID);
	UITextData* UIText_Get(unsigned int ID);
	bool UIText_Has(unsigned int ID);
	void UIText_SetText(unsigned int ID,const char* Text);
	void UIText_SetTextSize(unsigned int ID,float Scale);
	void UIText_SetFont(unsigned int ID,const char* FontName);
	void UIText_SetColor(unsigned int ID,float R,float G,float B,float A);
]]

---@class UITextData
local UITextData = {}

---@param Text string
function UITextData:SetText(Text)
	return dll.UIText_SetText(self.thisID,Text)
end

---@param Scale number
function UITextData:SetTextSize(Scale)
	return dll.UIText_SetTextSize(self.thisID,Scale)
end

---@param FontName string
function UITextData:SetFont(FontName)
	return dll.UIText_SetFont(self.thisID,FontName)
end

---@param R number
---@param G number
---@param B number
---@param A number
function UITextData:SetColor(R,G,B,A)
	return dll.UIText_SetColor(self.thisID,R,G,B,A)
end

local UIText_mt = {
	__index = UITextData
}

ffi.metatype("UITextData", UIText_mt)

UIText = UIText or {}

---@param ID number 
---@return UITextData 
function UIText.Add(ID) 
	local data = dll.UIText_Add(ID)
	data.thisID = ID
	return data 
end 

---@param ID number 
---@return UITextData 
function UIText.Get(ID) 
	return dll.UIText_Get(ID) 
end 

---@param ID number 
---@return boolean 
function UIText.Has(ID) 
	return dll.UIText_Has(ID) 
end 

return UIText